tests/queries/test_bulk_save.py

```diff
@@ -0,0 +1,210 @@
+import datetime
+
+from django.db.models import F
+from django.db.models.functions import Lower
+from django.test import TestCase
+
+from .models import (
+    Article, CustomDbColumn, CustomPk, Detail, Individual, Member, Note,
+    Number, Paragraph, SpecialCategory, Tag, Valid,
+)
+
+
+class BulkSaveTests(TestCase):
+    def setUp(self):
+        self.notes = [
+            Note.objects.create(note=str(i), misc=str(i))
+            for i in range(10)
+        ]
+        self.special_categories = [
+            SpecialCategory.objects.create(name=str(i), special_name=str(i))
+            for i in range(10)
+        ]
+        self.tags = [
+            Tag.objects.create(name=str(i))
+            for i in range(10)
+        ]
+
+    def test_simple(self):
+        for note in self.notes:
+            note.note = 'test-%s' % note.id
+        with self.assertNumQueries(1):
+            Note.objects.bulk_save(self.notes, ['note'])
+        self.assertCountEqual(
+            Note.objects.values_list('note', flat=True),
+            [cat.note for cat in self.notes]
+        )
+
+    def test_multiple_fields(self):
+        for note in self.notes:
+            note.note = 'test-%s' % note.id
+            note.misc = 'misc-%s' % note.id
+        with self.assertNumQueries(1):
+            Note.objects.bulk_save(self.notes, ['note', 'misc'])
+        self.assertCountEqual(
+            Note.objects.values_list('note', flat=True),
+            [cat.note for cat in self.notes]
+        )
+        self.assertCountEqual(
+            Note.objects.values_list('misc', flat=True),
+            [cat.misc for cat in self.notes]
+        )
+
+    def test_custom_pk(self):
+        custom_pks = [
+            CustomPk.objects.create(name='pk-%s' % i, extra='')
+            for i in range(10)
+        ]
+        for model in custom_pks:
+            model.extra = 'extra-%s' % model.pk
+        CustomPk.objects.bulk_save(custom_pks, ['extra'])
+        self.assertCountEqual(
+            CustomPk.objects.values_list('extra', flat=True),
+            [cat.extra for cat in custom_pks]
+        )
+
+    def test_inherited_fields(self):
+        for idx, category in enumerate(self.special_categories):
+            category.name = 'test-%s' % category.id
+            category.special_name = 'special-test-%s' % category.special_name
+        SpecialCategory.objects.bulk_save(self.special_categories, ['name', 'special_name'])
+        self.assertCountEqual(
+            SpecialCategory.objects.values_list('name', flat=True),
+            [cat.name for cat in self.special_categories]
+        )
+        self.assertCountEqual(
+            SpecialCategory.objects.values_list('special_name', flat=True),
+            [cat.special_name for cat in self.special_categories]
+        )
+
+    def test_update_primary_key(self):
+        for note in self.notes:
+            note.pk += 1
+        Note.objects.bulk_save(self.notes, ['id'])
```

Comment:
Is it intentional not to have any assertions?

Reply:
Nope, I must have forgotten that as well. I can add one this evening.

Reply:
I'm glad you flagged this because this test does not make any sense in retrospect. We cannot update the primary key using this method, so this needs to raise an exception and be documented.

We do a case statement based on the primary key value, so if you change that to an arbitrary value how did I expect the case statement to be valid or work at all?